aptos-move/framework/aptos-framework/sources/multisig_account.move

```diff
@@ -0,0 +1,1081 @@
+    /// to provide the full payload, which will be validated against the hash stored on-chain.
+    ///
+    /// @param function_hash The sha-256 hash of the function to invoke, e.g. 0x123::module_to_call::function_to_call.
+    /// @param args_hash The sha-256 hash of the function arguments - a concatenated vector of the bcs-encoded
+    /// function arguments.
+    public entry fun create_transaction_with_hash(
+        owner: &signer,
+        multisig_account: address,
+        function_hash: vector<u8>,
+        args_hash: vector<u8>,
+    ) acquires MultisigAccount {
+        // These are sha3-256 hashes so they must be exactly 32 bytes.
+        assert!(vector::length(&function_hash) == 32, error::invalid_argument(EINVALID_FUNCTION_HASH));
+        assert!(vector::length(&args_hash) == 32, error::invalid_argument(EINVALID_FUNCTION_ARGS_HASH));
+
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        assert_is_owner(owner, multisig_account_resource);
+
+        let creator = address_of(owner);
+        let transaction = MultisigTransaction {
+            payload: option::none<TransactionPayload>(),
+            payload_hash: option::some(PayloadHash { function_hash, args_hash }),
+            approvals: simple_map::create<address, bool>(),
+            rejections: simple_map::create<address, bool>(),
+            creator,
+        };
+        add_transaction(creator, multisig_account_resource, transaction);
+    }
+
+    /// Approve a multisig transaction.
+    public entry fun approve_transaction(
+        owner: &signer, multisig_account: address, transaction_id: u64) acquires MultisigAccount {
+        vote_transanction(owner, multisig_account, transaction_id, true);
+    }
+
+    /// Reject a multisig transaction.
+    public entry fun reject_transaction(
+        owner: &signer, multisig_account: address, transaction_id: u64) acquires MultisigAccount {
+        vote_transanction(owner, multisig_account, transaction_id, false);
+    }
+
+    /// Generic function that can be used to either approve or reject a multisig transaction
+    public entry fun vote_transanction(
+        owner: &signer, multisig_account: address, transaction_id: u64, approved: bool) acquires MultisigAccount {
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        assert_is_owner(owner, multisig_account_resource);
+
+        assert!(
+            table::contains(&multisig_account_resource.transactions, transaction_id),
+            error::not_found(ETRANSACTION_NOT_FOUND),
+        );
+        let transaction = table::borrow_mut(&mut multisig_account_resource.transactions, transaction_id);
+        let owner_addr = address_of(owner);
+        if (approved) {
```

Comment:
I think there is a potential bug here. Since you split the approvals and rejections into 2 simple maps, an address could vote for both. That breaks the assumption. 
I suggested to have one `votes` simple_map to replace those two simple maps so that `true/false` can make more sense.

Reply:
I don't think that's a bug. Yes an owner can vote for both but this is a multisig transaction and not really a broader voting flow so there shouldn't be contention on whether a transaction can be executed. With that said, we could still make it safe by revoking the owner's older vote. The reason for keeping two simple maps is to make it cheaper to execute a transaction (number of signatures can be tracked by looking at the map length instead of iterating through each entry).

Reply:
It's not possible now after the changes I made last time. When an owner changes their vote now, it nullifies their earlier vote.

Reply:
I read the new implementation. Then I think it is better to use a `simple_map<address, bool>` that includes all the votes, `true` means +, `false` means -. And you can have a counter to count the number of approval so far. So the add and remove will be simpler as simple_map will take care of the replacement.

Reply:
Your proposal would lead to netting of approvals against rejections though and technically change the count here. From my experience with some multisig wallets on other chain, the more common counting mechanism is to allow approval or rejection if the count of either reaches signatures threshold, not the net approvals (approvals - rejections).